### ISSUE TITLE
RangeError on Discover

### ISSUE THREAD

**Issue by PimButton** [sandeepmistry/noble#241]
I'm getting the RangeError when running the peripheral-explorer example. I'm on windows, using a fresh install of both Noble and Async. 
```
Your environment has been set up for using Node.js 0.12.7 (ia32) and npm.

C:\Users\Kokkerelletje\Documents\VanErum\PC Applicatie\testNoble>node test.js
buffer.js:506
    throw new RangeError('index out of range');
          ^
RangeError: index out of range
    at checkOffset (buffer.js:506:11)
    at Buffer.readUInt8 (buffer.js:544:5)
    at BluetoothHciSocket.onHciEventEndpointData (C:\Users\Kokkerelletje\Documents\VanErum\PC Applicatie\testNoble\node_modules\noble\node_modules\bluetooth-hci
-socket\lib\usb.js:115:36)
    at InEndpoint.emit (events.js:107:17)
    at Transfer.transferDone (C:\Users\Kokkerelletje\Documents\VanErum\PC Applicatie\testNoble\node_modules\noble\node_modules\bluetooth-hci-socket\node_modules
\usb\usb.js:295:9)
```

Any ideas? Would like to provide more debug data, how do I enable this? The error looks like #25, but that one is closed and I cannot reopen it.

**Comment by sandeepmistry** [comment]
@PimButton thanks for reporting, debug logs would be useful, along with you sharing your code and steps to reproduce.

```sh
set DEBUG=*
node <file>.js
```

**Comment by PimButton** [comment]
Ok, I have some more info. The device I am connecting to seems to be the problem. It's a custom designed PCB with a NRF51 as bluetooth device. Noble on the Mac and the Bluetooth Explorer on the Mac have no trouble connecting and writing / reading, so it seems to be a windows only problem. 

## The log
The debug log shows the following: 

```
C:\Users\Kokkerelletje\Documents\VanErum\PC Applicatie\testNoble>node test.js
  hci-usb reset +0ms
  hci-usb write: 01030c00 +15ms
  hci-usb HCI event: 0e04010c200c +16ms
  hci onSocketData: 040e04010c200c +15ms
  hci   event type = 4 +0ms
  hci   sub event type = 14 +16ms
  hci           cmd = 8204 +16ms
  hci           status = 12 +15ms
  hci           result =  +16ms
  hci-usb HCI event: 0e0401030c00 +31ms
  hci-usb reset complete +16ms
  hci onSocketData: 040e0401030c00 +15ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +0ms
  hci           cmd = 3075 +15ms
  hci           status = 0 +16ms
  hci           result =  +0ms
  hci setting filter to: 1400000020c10000000000400000 +811ms
  hci set event mask - writing: 01010c08fffffbff07f8bf3d +16ms
  hci-usb write: 01010c08fffffbff07f8bf3d +0ms
  hci set le event mask - writing: 010120081f00000000000000 +15ms
  hci-usb write: 010120081f00000000000000 +16ms
  hci read local version - writing: 01011000 +16ms
  hci-usb write: 01011000 +15ms
  hci read bd addr - writing: 01091000 +0ms
  hci-usb write: 01091000 +16ms
  hci-usb HCI event: 0e0401010c00 +15ms
  hci onSocketData: 040e0401010c00 +16ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +0ms
  hci           cmd = 3073 +15ms
  hci           status = 0 +16ms
  hci           result =  +15ms
  hci-usb HCI event: 0e0401012000 +0ms
  hci onSocketData: 040e0401012000 +16ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +0ms
  hci           cmd = 8193 +15ms
  hci           status = 0 +16ms
  hci           result =  +15ms
  hci-usb HCI event: 0e0c0101100006bb22060a00bb22 +0ms
  hci onSocketData: 040e0c0101100006bb22060a00bb22 +16ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +15ms
  hci           cmd = 4097 +0ms
  hci           status = 0 +16ms
  hci           result = 06bb22060a00bb22 +15ms
  hci set scan enabled - writing: 010c20020001 +16ms
  hci-usb write: 010c20020001 +16ms
  hci set scan parameters - writing: 010b200701100010000000 +0ms
  hci-usb write: 010b200701100010000000 +15ms
  hci-usb HCI event: 0e0a010910001371da7d1a00 +16ms
  hci onSocketData: 040e0a010910001371da7d1a00 +15ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +0ms
  hci           cmd = 4105 +16ms
  hci           status = 0 +15ms
  hci           result = 1371da7d1a00 +16ms
  hci address = 00:1a:7d:da:71:13 +0ms
  hci-usb HCI event: 0e04010c200c +15ms
  hci onSocketData: 040e04010c200c +16ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +0ms
  hci           cmd = 8204 +15ms
  hci           status = 12 +16ms
  hci           result =  +15ms
  hci-usb HCI event: 0e04010b2000 +16ms
  hci onSocketData: 040e04010b2000 +0ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +15ms
  hci           cmd = 8203 +16ms
  hci           status = 0 +0ms
  hci           result =  +15ms
  noble stateChange poweredOn +16ms
Started scanning
  hci set scan enabled - writing: 010c20020101 +31ms
  hci-usb write: 010c20020101 +0ms
  hci-usb HCI event: 0e04010c2000 +16ms
  hci onSocketData: 040e04010c2000 +15ms
  hci   event type = 4 +16ms
  hci   sub event type = 14 +16ms
  hci           cmd = 8204 +0ms
  hci           status = 0 +15ms
  hci           result =  +16ms
  noble scanStart +15ms
  hci-usb HCI event: 3e1c02010001ad5a4b49b1dc100c0969 +0ms
  hci-usb HCI event: 4d4f2d74657374202331020105c8 +16ms
  hci onSocketData: 043e1c02010001ad5a4b49b1dc100c09694d4f2d74657374202331020105c8 +16ms
  hci   event type = 4 +0ms
  hci   sub event type = 62 +15ms
  hci           LE meta event type = 2 +16ms
  hci           LE meta event status = 1 +15ms
  hci           LE meta event data = 0001ad5a4b49b1dc100c09694d4f2d74657374202331020105c8 +0ms
  hci                   type = 0 +16ms
  hci                   address = dc:b1:49:4b:5a:ad +16ms
  hci                   address type = random +15ms
  hci                   eir = 0c09694d4f2d74657374202331020105 +0ms
  hci                   rssi = -56 +16ms
  gap advertisement = {"localName":"iMO-test #1","serviceData":[],"serviceUuids":[]} +15ms
  hci-usb HCI event: 3e1e02010401ad5a4b49b1dc1211073b +16ms
  hci-usb HCI event: 1f465aadc949c69c2b13710100b38bc5 +16ms
  hci onSocketData: 043e1e02010401ad5a4b49b1dc1211073b1f465aadc949c69c2b13710100b38bc5 +15ms
  hci   event type = 4 +0ms
  hci   sub event type = 62 +16ms
  hci           LE meta event type = 2 +15ms
  hci           LE meta event status = 1 +16ms
  hci           LE meta event data = 0401ad5a4b49b1dc1211073b1f465aadc949c69c2b13710100b38bc5 +0ms
  hci                   type = 4 +16ms
  hci                   address = dc:b1:49:4b:5a:ad +15ms
  hci                   address type = random +16ms
  hci                   eir = 11073b1f465aadc949c69c2b13710100b38b +0ms
  hci                   rssi = -59 +15ms
  gap advertisement = {"localName":"iMO-test #1","serviceData":[],"serviceUuids":["8bb3000171132b9cc649c9ad5a461f3b"]} +16ms
Found something (iMO-test #1)
  hci set scan enabled - writing: 010c20020001 +31ms
  hci-usb write: 010c20020001 +0ms
peripheral with ID undefined found
  Local Name        = iMO-test #1
  Service Data      =
  Service UUIDs     = 8bb3000171132b9cc649c9ad5a461f3b

services and characteristics:
  hci create le conn - writing: 010d2019600030000001ad5a4b49b1dc0006000c000000c80004000600 +94ms
  hci-usb write: 010d2019600030000001ad5a4b49b1dc0006000c000000c80004000600 +0ms
  hci-usb HCI event:  +15ms
  hci set scan enabled - writing: 010c20020001 +16ms
  hci-usb write: 010c20020001 +16ms
buffer.js:506
    throw new RangeError('index out of range');
          ^
RangeError: index out of range
    at checkOffset (buffer.js:506:11)
    at Buffer.readUInt8 (buffer.js:544:5)
    at BluetoothHciSocket.onHciEventEndpointData (C:\Users\Kokkerelletje\Documents\VanErum\PC Applicatie\testNoble\node_modules\noble\node_modules\bluetooth-hci
-socket\lib\usb.js:115:36)
    at InEndpoint.emit (events.js:107:17)
    at Transfer.transferDone (C:\Users\Kokkerelletje\Documents\VanErum\PC Applicatie\testNoble\node_modules\noble\node_modules\bluetooth-hci-socket\node_modules
\usb\usb.js:295:9)
```

## The code
The code I'm using is a slightly adjusted version of the peripheral-explorer.js that I found in example. I have uploaded my version here: http://pastebin.com/3HC5pFGZ

## Steps to reproduce:
I'm using Windows 7 (64bit), Node 0.12.7 and a USB bluetooth dongle with the CSR8510 A10 chipset. Just run ```node testNoble.js``` in order to reproduce.

**Comment by sandeepmistry** [comment]
@PimButton thanks.

Woah, that's really strange, an empty HCI event is received here: https://github.com/sandeepmistry/node-bluetooth-hci-socket/blob/master/lib/usb.js#L101

```
hci-usb HCI event:  +15ms
```

Could you please try changing this function, https://github.com/sandeepmistry/node-bluetooth-hci-socket/blob/master/lib/usb.js#L100 to:

```javascript
BluetoothHciSocket.prototype.onHciEventEndpointData = function(data) {
  debug('HCI event: ' + data.toString('hex'));

  if (data.length === 0) {
    return;
  }  

  // ...
};
```

and then run again to see if things go further.

**Comment by PimButton** [comment]
@sandeepmistry  Yes, that fixed it, thanks! Now running into something else, will open a new issue.

**Comment by sandeepmistry** [comment]
@PimButton great! I've pushed the suggested change: https://github.com/sandeepmistry/node-bluetooth-hci-socket/commit/f2b05b724fa80adca22ed4d4c4817742016b1358 and also published v0.3.4 to npm.
